### ISSUE TITLE
2.0 - resolve LATEST-EAP-SNAPSHOT

### ISSUE THREAD

**Issue by jonathanlermitage** [JetBrains/intellij-platform-gradle-plugin#1628]
### Describe the need of your request

With gradle plugin v1, I was using intellij `{ version.set("IC-LATEST-EAP-SNAPSHOT") }`
With gradle plugin v2, I am using dependencies `{ intellijPlatform { intellijIdeaCommunity("...") } }`
but, what should I use as intellijIdeaCommunity's parameter? I tried LATEST-EAP-SNAPSHOT but it doesn't work.
I am looking at https://plugins.jetbrains.com/docs/intellij/tools-intellij-platform-gradle-plugin-faq.html#how-to-check-the-latest-available-eap-release -> should I really compute the latest EAP snapshot version string by myself? Is there a helper for that? Thx.

### Proposed solution

Allow `{ intellijPlatform { intellijIdeaCommunity("LATEST-EAP-SNAPSHOT") } }` or something similar, as in plugin v1.

**Comment by YannCebron** [comment]
Note, in general building against LATEST-EAP-SNAPSHOT is not recommended, as it can have unpredictable results (and you'll usually want to target a specific minimum EAP build).

**Comment by YannCebron** [comment]
There's no need to _build_ against some EAP version,  it's sufficient to configure it to run against that EAP build.

**Comment by jonathanlermitage** [comment]
To give more context, I used to test my plugin (by testing, I mean running `gradlew runIde`) with:
- a minimal IDE version. This version never changes, so it's har-coded. This is what I am using in my `since-build` property.
- the latest stable IDE version. I have a script that finds this version string. I'm doing this because I don't want a `until-build` property. I don't want to publish new releases just for updating this tag. But, this is another story.
- the latest EAP. I explained why.

**Comment by hsz** [comment]
This is already possible. To build your plugin with a specific IntelliJ Platform but run with another, for testing purposes, you have to define a custom task with:
```
val runLatestEapIde by tasks.registering(CustomRunIdeTask::class) {
    version = "LATEST-EAP-SNAPSHOT"
}
```

Note that, by default, we use now JetBrains CDN where only released IDEs are hosted.
To switch to IJ Maven Repository where snapshots are present, add this line to the `gradle.properties`.

```
org.jetbrains.intellij.platform.buildFeature.useBinaryReleases=false
```

This property name as well as the CDN/Maven resolution may soon change.

**Comment by bric3** [comment]
Since 2.0.0 is out the property mentioned above is gone, one can use something like this.

```
intellijPlatformTesting {
    runIde {
        create("runIntelliJLatest") {
            type = IntelliJPlatformType.IntellijIdeaCommunity
            useInstaller = false
            val versionProvider = providers.gradleProperty("runPlatformLatestVersion").orElse("LATEST-EAP-SNAPSHOT")
            version = versionProvider
            sandboxDirectory = providers.zip(layout.buildDirectory.dir("idea-sandbox/runIntelliJ"), versionProvider) { dir, version ->
                dir.dir(version)
            }
        }
    }
}
```

The only issue I found is that the JBR is not used (if other dependencies are not using it), e.g. here the main version don't use the installer and as such dont need the JBR, but adding the JBR runtime dependency to help the `runIntelliJLatest` don't seem to work for that task:

```
dependencies
    intellijPlatform {
        intellijIdeaCommunity(project.providers.gradleProperty("buildPlatformVersion"))

        pluginVerifier()
        instrumentationTools()
        zipSigner()

        jetbrainsRuntime()
    }
}
```

**Comment by jonathanlermitage** [comment]
@bric3 I find the latest EAP snapshot from `https://www.jetbrains.com/updates/updates.xml` by using an xpath expression: `/products/product[@name='IntelliJ IDEA']/channel[@id='IC-IU-EAP-licensing-EAP']/build[1]/@fullNumber`. It returns something like `243.12818.47`.  
I'm using it since a while, and it always refers to IDE builds with a JBR. It may skip very few EAP builds (I did not check), but at least it works.

**Comment by bric3** [comment]
@jonathanlermitage Thank you ! Looks interesting.

**Comment by lppedd** [comment]
I'm late to the party, but the overall build script configuration has become so overly complex I don't even want to make the effort of looking into it anymore.

If I want to use the latest EAP, I should just be able to use the placeholder as in v1 and that's it.

**Comment by joshuataylor** [comment]
Yeah, it's really hairy to do this, usecase is we always want to get the latest EAP snapshot without having to always version bump every few months to the latest 253.xx or 261.xxx etc.

And we want to test almost every IDE, so the build scripts are hairy.

**Comment by jonathanlermitage** [comment]
For info, I implemented a [scrapper]([url](https://github.com/jonathanlermitage/software-updates-bot)). The latest EAP build number is available here: https://github.com/jonathanlermitage/software-updates-bot/blob/master/report/uc.properties#L6

It's updated every 4 hrs. You can use these build numbers when using the `installer` (i.e. they have the JBR).

**Comment by hsz** [comment]
I just got an idea — we already have an internal mechanism for resolving installer IntelliJ Platform distributions available in CDN, which is utilized in Plugin Verifier's configuration:

```kotlin
intellijPlatform {
    pluginVerification {
        ides {
            recommended()
            
            select {
                sinceBuild = "251"
                untilBuild = "253.*"
                types = listOf(IntelliJPlatformType.IntellijIdea)
                channels = listOf(ProductRelease.Channel.RELEASE)
            }
        }
    }
}
```

Both `select {}` and `recommended()` helpers use the `ProductReleasesValueSource()` under the hood, which requests data from

```kotlin
const val PRODUCTS_RELEASES_ANDROID_STUDIO = "https://jb.gg/android-studio-releases-list.xml"
const val PRODUCTS_RELEASES_JETBRAINS_IDES = "https://www.jetbrains.com/updates/updates.xml"
```

Note: the `recommended()` uses `select {}` and prefills it with the type, since, until, and channels set to "release, eap, rc".

I successfully moved this `ProductReleasesValueSource()` mechanism upper so it can be available via the root `intellijPlatform` extension, like:

```kotlin
val latestIUProvider = intellijPlatform.ProductReleasesValueSource().map { it.first() }

dependencies {
    intellijPlatform {
        create(latestIUProvider)
    }
}
```

Here, we request all available IUs and get the first, latest one. This value source provider returns a set of IDE notation strings that work well with the `create()` helper.

It's still possible to work on the `version` only by calling the `String.parseIdeNotation() : Pair<IntelliJPlatformType, String>` helper (available publicly since 2.11.0).

```kotlin
val latestIUProvider = intellijPlatform.ProductReleasesValueSource().map {
    it.first().parseIdeNotation().second
}

dependencies {
    intellijPlatform {
        intellijIdea(latestIUProvider)
    }
}
```

Of course, this helper takes extra params: 

```kotlin
fun ProductReleasesValueSource(configure: FilterParameters.() -> Unit = {})
```

see: https://github.com/JetBrains/intellij-platform-gradle-plugin/blob/1157bc0b5f1749c5f02e0fe79ddc6a7404c0c52e/src/main/kotlin/org/jetbrains/intellij/platform/gradle/providers/ProductReleasesValueSource.kt#L55-L83

WDYT?

**Comment by hsz** [comment]
Thanks for the report — this is now supported directly.

Installer-based IntelliJ Platform dependencies (and custom testing configurations) now accept `"latest"` (or `Constraints.LATEST_VERSION`) as the version. It resolves to the newest available release for the requested platform type **across all channels** (backed by `ProductReleasesService`), using the installer path (`useInstaller = true`, the default):

```kotlin
intellijPlatform {
    intellijIdea("latest")
    // or
    intellijIdea(Constraints.LATEST_VERSION)
}
```

It also works for custom testing tasks registered via `intellijPlatformTesting`:

```kotlin
val runLatestIde = intellijPlatformTesting.runIde.register("runLatestIde") {
    type = IntelliJPlatformType.IntellijIdea
    version = "latest"
}
```

A couple of clarifications:

- **`LATEST-EAP-SNAPSHOT` is still supported** for the non-installer path (`useInstaller = false`, i.e. Maven snapshots / Multi-OS ZIP, typically together with `snapshots()`). The new `"latest"` keyword does **not** replace it — it's the installer-based counterpart.
- Tracking `"latest"` means your platform version is not pinned and therefore not fully reproducible build-to-build. If you need a fixed version, the `printProductsReleases` task still lists the matching releases so you can pin one by hand.

Docs (available after the next release):
- Target versions / `latest`: https://plugins.jetbrains.com/docs/intellij/tools-intellij-platform-gradle-plugin-dependencies-extension.html#target-versions-latest
- FAQ — How to check the latest available EAP release: https://plugins.jetbrains.com/docs/intellij/tools-intellij-platform-gradle-plugin-faq.html#how-to-check-the-latest-available-eap-release

**Comment by joshuataylor** [comment]
Amazing, just in time for the new EAPs! :tada:
